<div style="width: 100%; clear: both;">
<div style="float: left; width: 50%;">
<img src="http://www.uoc.edu/portal/_resources/common/imatges/marca_UOC/UOC_Masterbrand.jpg", align="left">
</div>
<div style="float: right; width: 50%;">
<p style="margin: 0; padding-top: 22px; text-align:right;">M2.975 · Deep Learning · PAC3</p>
<p style="margin: 0; text-align:right;">2024-2 · Màster universitari en Ciència de dades (Data science)</p>
<p style="margin: 0; text-align:right; padding-button: 100px;">Estudis d'Informàtica, Multimèdia i Telecomunicació</p>
</div>
</div>
<div style="width:100%;">&nbsp;</div>


# PAC 3: Transformers amb Keras


<u>Consideracions generals</u>:

- Aquesta PAC s'ha de realitzar de forma **estrictament individual**. Qualsevol indicatiu de còpia serà penalitzat amb un suspens (D) per a totes les parts implicades i la possible avaluació negativa de l'assignatura de manera íntegra.
- És necessari que l'estudiant indiqui **totes les fonts** que ha utilitzat per a la realització de la PAC. De no ser així, es considerarà que l'estudiant ha comès plagi, sent penalitzat amb un suspens (D) i la possible avaluació negativa de l'assignatura de manera íntegra.

<u>Format de l'entrega</u>:

- Alguns exercicis poden suposar diversos minuts d'execució, per la qual cosa l'entrega s'ha de fer en **format notebook** i en **format html**, on es vegi el codi, els resultats i comentaris de cada exercici. Es pot exportar el notebook a HTML des del menú File $\to$ Download as $\to$ HTML.
- Existeix un tipus de cel·la especial per allotjar text. Aquest tipus de cel·la us serà molt útil per respondre a les diferents preguntes teòriques plantejades al llarg de l'activitat. Per canviar el tipus de cel·la a aquest tipus, en el menú: Cell $\to$ Cell Type $\to$ Markdown.

## Càrrega de llibreries
A continuació es mostren les principals llibreries necessàries per executar aquesta PAC.

In [ ]:
import tensorflow as tf
import keras
from transformers import BertTokenizer, TFBertModel

from tensorflow.keras import layers, models
from tensorflow.keras.layers import Input, Dense, Embedding, GlobalAveragePooling1D
from keras.layers import LSTM, SimpleRNNCell, Flatten, Dropout, MultiHeadAttention, Bidirectional, TimeDistributed, Dense, Embedding
from tensorflow.keras.layers import Permute, dot, add, concatenate
from tensorflow.keras.utils import plot_model
from tensorflow.keras.callbacks import ModelCheckpoint,EarlyStopping,ReduceLROnPlateau


from tensorflow.keras.preprocessing.sequence import pad_sequences

import tensorflow_hub as hub
import tensorflow_text as text
from official.nlp import optimization

import matplotlib.pyplot as plt



2025-04-16 16:39:58.981621: I tensorflow/core/util/port.cc:111] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-04-16 16:39:58.984374: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2025-04-16 16:39:59.028870: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2025-04-16 16:39:59.028919: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2025-04-16 16:39:59.028952: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to regi

In [ ]:
print(tf.__version__)
print(hub.__version__)
print(text.__version__)

2.14.0
0.16.1
2.14.0


<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Es recomana en aquesta PAC utilitzar tensorflow 2.14. Utilitzar altres versions pot dificultar el treball en la part sobre downstream models a causa de la incompatibilitat de llibreries</strong> </p>
</div>

## Context

En aquesta PAC utilitzarem models de transformers, capes d'atenció, capes bidireccionals i alguns altres algorismes estudiats fins ara per construir i entrenar fent dos exercicis relacionats amb el llenguatge natural. Aquest entorn és on les arquitectures basades en transformers són més potents.


Els objectius de la pràctica són:

* Càrrega, preprocessament, exploració i visualització de les dades.
* Entrenament d'un model base amb els coneixements adquirits fins ara.
* Comprensió de les capes bidireccionals.
* Comprensió i aplicació de les capes d'atenció.
* Comprensió i aplicació de blocs transformers bàsics.
* Comprensió i aplicació de blocs transformers avançats.
* Aplicació sobre domini (downstream models) d'arquitectures preentrenades basades en transformers.
* Adquisició de coneixements i capacitat crítica que permeti continuar millorant els resultats.

## 1 Càrrega de dades (0.5 punts)


<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [0.5 pt]:</strong> IMBD és una base de dades que conté ressenyes de pel·lícules, juntament amb la valoració sobre si el comentari és positiu o negatiu. En aquesta primera part realitza el següent:

* Descarrega la base de dades d'IMBD des de keras i mostra 2 exemples de comentaris positius i 2 exemples de comentaris negatius. Les paraules estan codificades en números, mostra els exemples en text. Tens informació de com fer-ho a: https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb/get_word_index
* Reprodueix la taula 1 (només les dades de la fila 3) del paper Long Short-Term Memory with Dynamic Skip Connections (https://arxiv.org/pdf/1811.03873v1). Calcula els valors d'aquesta taula sobre la base de dades de l'IMBD amb què estem treballant. Fes particions de dades o les suposicions que consideris. Si hi hagués alguna diferència amb el paper, indica-la.
</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 2 Model base LSTM (1 punt)


<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [1 pt]:</strong>

    
Ara intentarem completar algunes parts de la taula 4 del paper Long Short-Term Memory with Dynamic Skip Connections https://arxiv.org/pdf/1811.03873v1.

Es recomana, en primer lloc, entendre les dificultats que afronta el llenguatge natural. Per això observa la Figura 1 del paper i llegeix la introducció, en particular el penúltim paràgraf.


* En aquesta part, entrena un model simple LSTM tal com apareix a la primera fila de la taula 4. Mostra les gràfiques d'accuracy i loss durant l'entrenament (de la partició d'eval i de train), així com el resultat final sobre test. Repasseu el text corresponent a la secció "Sentiment Analysis on IMDB" per intentar reproduir els resultats. Discuteix les decisions que has hagut de prendre en cas de no tenir tota la informació. Implementa un early stopping i queda't sempre amb el millor model sobre el dataset d'avaluació.
    
</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 3 Model basat en mecanismes d'atenció. (2 punts)



<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [2 pts.]:</strong>
En aquest exercici, hauràs de construir i entrenar una petita arquitectura que combini una capa d'Embedding, mecanismes d'atenció (Multi-Head Attention) i connexions residuals, simulant els principis bàsics dels Transformers. Per això el model constarà:


* Entrada del model.
* Capa d'Embedding: Utilitza una capa Embedding.
* Capa d'Atenció: Incorpora una capa MultiHeadAttention. Aquesta capa ha de rebre com a entrada la seva pròpia sortida, simulant una atenció self-attention.
* Connexió Residual: Afegeix la sortida de la capa d'atenció a la sortida original de la capa d'embedding mitjançant una operació de suma (Add). Posteriorment, aplica una capa de normalització (LayerNormalization).
* Capa de Flatten i Regularització.
* Finalment afegeix una darrera capa classificadora d'una neurona i activació sigmoide

Amb aquesta arquitectura hauràs de:
* Mostrar el resum del model i el diagrama de blocs.
* Entrenar el model i obtenir els resultats sobre test. Mostra les corbes d'entrenament. Quan entrenis el model, també implementa un early stopping i queda't amb el model que et doni millors resultats en validació.
* Comenta i discuteix què has realitzat. Compara els teus resultats amb els obtinguts anteriorment. Amplia la taula de resultats demanada a l'exercici anterior. Pren les decisions sobre els paràmetres que consideris.

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 4 Model basat en mecanismes d'atenció millorat. (0.5 punts)



<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [0.5 pts.]:</strong>
Fes alguna modificació sobre l'arquitectura anterior i reentrena el model. Algunes idees poden ser: afegir diversos caps d'atenció, modificar els residus, incloure-hi alguna capa recurrent, etc.

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 5 Downstream model. (3 punts)
Un model downstream (o tasca downstream) és un model o tasca que utilitza un model preentrenat en una fase anterior, sovint coneguda com a fase de preentrenament, per resoldre una tasca específica. Aquest model posteriorment es fa servir mitjançant un entrenament específic per resoldre una tasca diferent. Si la tasca té característiques similars i l'arquitectura de la xarxa és prou complexa com per captar els fonaments compartits, és una estratègia que porta molts avantatges.

Un dels grans avantatges dels Transformers és la seva escalabilitat. Per entendre això es recomana ampliar els coneixements amb alguns recursos divulgatius com els següents:
* https://www.youtube.com/watch?v=EFkbT-1VGTQ
* https://www.fabricatedknowledge.com/p/ai-foundations-part-1-transformers
* https://www.baeldung.com/cs/rnns-transformers-nlp

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [2 pt]:</strong>

En aquest exercici prendrem les dades que hem utilitzat fins ara i farem un downstream model del model BERT. Aquesta part està mostrada al següent tutorial: https://www.tensorflow.org/text/tutorials/classify_text_with_bert

Hauràs d'adaptar-la per treballar amb les dades que hem carregat anteriorment. Tria la versió del model que consideris i pren les decisions adequades justificant-les posteriorment. La dificultat principal en aquest exercici és adaptar el codi del tutorial a les dades i particions que has carregat durant l'exercici. Per això:
* Entrena un model amb l'arquitectura BERT més petita mostrada al tutorial anterior. Trieu adequadament el nombre d'èpoques i controleu l'entrenament amb un early stopping
* Mostra el resum del model amb el nombre de paràmetres
* Mostra les gràfiques d'entrenament i validació.
* Finalment completa la taula de resultats i mostra-la

</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solución:</strong> </p>
</div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [1 pt]:</strong>

Repeteix l'exercici anterior però ara fent servir un model BERT més gran. Heu d'utilitzar un model que sigui capaç d'obtenir millors resultats que en tots els casos anteriors. Finetunejar un model gran pot requerir diverses hores.
* Afegeix a la taula de resultats, el resultat obtingut amb aquest model.
* Igual que abans representa les gràfiques d'accuracy i loss.
* Comenta els resultats i discuteix els temps d'entrenament emprats

</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 6 Optimitzant els entrenaments. (2 punts)
Hauràs observat que un dels desavantatges de les grans arquitectures transformers de xarxes neuronals és que requereixen una gran potència de còmput per entrenar-se correctament. Això no és només un problema de temps, sinó de diners.
El maquinari i l'electricitat suposen el cost operatiu principal d'aquestes arquitectures, per la qual cosa optimitzar-les és primordial i suposa un avantatge competitiu (vegeu el cas de DeepSeek).

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [0.5 pt]:</strong>

Repeteix l'exercici anterior però ara només prova les últimes capes. És a dir, en aquest cas el bloc BERT es queda congelat. Podeu provar d'afegir alguna capa entre el bloc BERT i la sortida.
* Afegeix a la taula de resultats, el resultat obtingut amb aquest model
* Comenta els resultats i discuteix els temps d'entrenament emprats. És una estratègia útil? Quines altres alternatives similars hi ha?

</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [1.5 pt]:</strong>  Catastrophic forgetting.

A la secció 5.3.3 del paper How to Fine-Tune BERT for Text Classification? https://arxiv.org/pdf/1905.05583v3 s'estudia el Catastrophic forgetting. Aquesta situació és molt delicada en els models de transformers de gran escala i s'ha de comprendre molt bé per evitar aquestes situacions

* Reprodueix la figura "Figure 2: Catastrophic Forgetting" d'aquest article científic. En concret, les subfigures c i d. Només cal que ho facis sobre la corba de train. L'eix x d'aquestes figures (iterations), pots considerar que són els batches d'entrenament, mentre que l'eix Y fa referència al loss. No cal que utilitzis exactament els valors de lr que es mostren en aquesta figura. L'objectiu és que mostris un cas en què el lr és tan gran que es produeix el Catastrophic forgetting, i un altre cas amb un altre valor en què això no passa.
</div>

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>

## 7 Estratègia final. (1 punt)

<div style="background-color: #EDF7FF; border-color: #7C9DBF; border-left: 5px solid #7C9DBF; padding: 0.5em;">
<p><strong>Exercici [1 pt]:</strong>
Amb tot allò que has après. Finalment, entrena un model de BERT per a classificació i intenta obtenir els millors resultats possibles.

<div style="background-color: #fcf2f2; border-color: #dfb5b4; border-left: 5px solid #dfb5b4; padding: 0.5em;">
<p><strong>Solució:</strong> </p>
</div>